app/Http/Controllers/Rpc/WalletController.php

```diff
@@ -0,0 +1,139 @@
+    const VALIDATOR_RULE_REQUIRED = 'required';
+
+    public function calculateWithdrawal(Request $request)
+    {
+        Validator::make($request->all(), [
+            self::FIELD_AMOUNT => [self::VALIDATOR_RULE_REQUIRED, 'integer', 'min:1'],
+            self::FIELD_TO => self::VALIDATOR_RULE_REQUIRED,
+        ])->validate();
+        $amount = $request->input(self::FIELD_AMOUNT);
+        $addressTo = $request->input(self::FIELD_TO);
+
+        $addressFrom = $this->getAdserverAdsAddress();
+        if (!AdsValidator::isAccountAddressValid($addressFrom)
+            || !AdsValidator::isAccountAddressValid($addressTo)) {
+            // invalid input for calculating fee
+            return self::json([], 422);
+        }
+        $fee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
+
+        $total = $amount + $fee;
+        $resp = [
+            self::FIELD_AMOUNT => $amount,
+            'fee' => $fee,
+            'total' => $total,
+        ];
+
+        return self::json($resp);
+    }
+
+    public function withdraw(Request $request)
+    {
+        Validator::make($request->all(), [
+            self::FIELD_AMOUNT => [self::VALIDATOR_RULE_REQUIRED, 'integer', 'min:1'],
+            self::FIELD_TO => self::VALIDATOR_RULE_REQUIRED,
+        ])->validate();
+
+        $amount = $request->input(self::FIELD_AMOUNT);
+        $addressTo = $request->input(self::FIELD_TO);
+        $memo = $request->input(self::FIELD_MEMO);
+
+        $addressFrom = $this->getAdserverAdsAddress();
+        if (!AdsValidator::isAccountAddressValid($addressFrom)
+            || !AdsValidator::isAccountAddressValid($addressTo)) {
+            // invalid input for calculating fee
+            return self::json([], 422);
+        }
+        $fee = AdsUtils::calculateFee($addressFrom, $addressTo, $amount);
+
+        $total = $amount + $fee;
+        if (!$this->hasUserEnoughFunds($total)) {
+            return self::json(['error' => 'not enough funds'], 400);
+        }
+
+        // TODO add tx to queue: $amount is amount, $addressTo is address
+
+        return self::json([], 204);
+    }
+
+    public function depositInfo()
+    {
+        $user = Auth::user();
+        $uuid = $user->uuid;
+        /**
+         * Address of account on which funds should be deposit
+         */
+        $address = $this->getAdserverAdsAddress();
+        /**
+         * Message which should be add to send_one tx
+         */
+        $message = str_pad($uuid, 64, '0', STR_PAD_LEFT);
+        $resp = [
+            self::FIELD_ADDRESS => $address,
+            self::FIELD_MESSAGE => $message,
+        ];
+
+        return self::json($resp);
+    }
+
+    /**
+     * Checks, if user has enough funds.
+     *
+     * @param $amount int transfer total amount
+     *
+     * @return bool true if has enough, false otherwise
+     */
+    private function hasUserEnoughFunds(int $amount): bool
+    {
+        // TODO check user account balance
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=php%3AS1135)